### ISSUE TITLE
Advise file system developers not to use file handles in nonportable ways

### ISSUE THREAD

**Issue by asomers** [libfuse/libfuse#389]
# Background
The FUSE protocol allows a server to assign a different filehandle each and every time that a file is opened.  Multiple filehandles can be concurrently valid for the same file.  Operations like `FUSE_READ` and `FUSE_WRITE` are supposed to include the file handle associated with the particular file descriptor that initiated the operation.  However, that isn't possible in all cases.  The `FUSE_WRITE_CACHE` bit indicates whether the file handle is valid for a `FUSE_WRITE` operation, and the `FATTR_FH` flag indicates whether it's valid for a `FUSE_SETATTR` operation.

# Problem
There are other cases, too, when the file handle can be invalid.  In fact, the entire concept of a file handle is really a Linuxism.  Other operating systems enforce a distinction between the file layer and the vnode layer.  That's why OSX, FreeBSD, OpenBSD, and Illumos can almost *never* guarantee that they're sending the right file handle for any operation.  NetBSD takes a more extreme position; it doesn't allow multiple concurrent fuse file handles for a single file (which leads to other problems).   But since operations other than `FUSE_WRITE` and `FUSE_SETATTR` don't have anything analogous to the `FUSE_WRITE_CACHE` or `FATTR_FH` bits, there's no way for those operating systems to indicate that the file handle is guessed.

# Mitigating Circumstances
The impact of this bug is small simply because few if any fuse servers care about the accuracy of fuse file handles.  I surveyed 42 real-world FUSE filesystems* and _none_ of them ever check the `writepage` bit.  Only _one_ of them (catfs) checks `FATTR_FH`, and it can deal with the absence of that flag.  I can't find any evidence that any FUSE filesystem actually wants to assign distinct file handles on each concurrent `FUSE_OPEN` and requires the FUSE client to keep them straight.

# Solution
Ideally the FUSE protocol would require servers to assign the same file handle for every concurrent open of a single file.  Unfortunately, that's not the current situation and such a change would be backwards-incompatible.  The second best solution would be to update the kernel protocol to include some equivalent of the `FUSE_WRITE_CACHE` bit for every operation that includes a file handle.  However, that would require all clients and servers alike to update their kernel API version, so uptake would be extremely slow.  The path of least resistance would simply be to document the problem.  I recommend the following:

* Advise FUSE server developers to assign the same file handle for every concurrent open of a single file.
* Rename the `writepage` bit to `fh_invalid`, and add a `#define` for backwards-compatibility
* Patch libfuse to unconditionally set `fh_invalid` for most commands on non-Linux operating systems.

If the maintainers agree, I'll prepare a PR.

cc @cemeyer @kendmerry

*: afuse, chironfs, cryptofs, curlftpfs, encfs, ext2, ext4fuse, funionfs, fusepak, gitfs, gstfs, gunzip, hfsfuse, httpfs, ifuse, ltfs, mhddfs, mp3fs, ntfs, ntfs-compression, pod, rar2fs, s3backer, s3fs, simple-mtpfs, smbnetfs, sqlfs, squashfuse, sshfs, unionfs, wdfs, webdavfs, wikipediafs, zip, MooseFS, CephFS, RedoxFS, gcsf, catfs, qcow2-fuse, shotwellvfs, and one closed-source file system.

**Comment by Nikratio** [comment]
Thanks for the detailed report!

When you say

> Operations like FUSE_READ and FUSE_WRITE are supposed to include the file handle associated with the particular file descriptor that initiated the operation

What do you mean with "supposed"? AFAIK the libfuse documentation makes no such promise. Neither documentation nor the FUSE protocol have a notion of file descriptors at all. 

This also wouldn't be possible in practice, since read-ahead and writeback caching break any 1:1 mapping between read/write request and file descriptors.

**Comment by Nikratio** [comment]
What I mean with the above is that it seems to me that filesystem developers already have no reason to make any assumptions about which file handle is passed in, and that *writepage* has well defined semantics that are unrelated to "correctness" of the file handle. We could advise to re-use file handles, but based on your description it's not clear to me why this would be advisable..

**Comment by asomers** [comment]
"File descriptor" is not precisely the right term, but libfuse doesn't define the right term.  What I'm referring to are concurrent opens of the same nodeid.  While libfuse doesn't specify when the client should concurrently open the same nodeid multiple times, it's my impression that when `-o default_permissions` is not used, the client must FUSE_OPEN a nodeid for each and every `open(2)` call.  Otherwise nothing would validate the permissions of the `open(2)`.  So what's the correct term for each concurrent open of a nodeid?

Libfuse's documentation for `release` says this ("file descriptor" should probably say "file handle" instead):
```
* For every open() call there will be exactly one release() call
* with the same flags and file descriptor.
```
If "file descriptor" is supposed to be "file handle", then that clearly requires the client to track all file handles ever returned by `open`.  And why would it track all of them if they were all required to be the same?  Furthermore, the documentation for the `FUSE_WRITE_CACHE` bit says:
```
FUSE_WRITE_CACHE: delayed write from page cache, file handle is guessed
```
If the file handle is guessed when that bit is set, it implies that the file handle must not be guessed when that bit is cleared.  And since `writepage` is set if and only if the client sends `FUSE_WRITE_CACHE`, the semantics of `writepage` most definitely *are* related to the correctness of the file handle.

From all of the above, it is pretty clear to me that for operations on an open file, the FUSE client must send the specific file handle returned by the server during `FUSE_OPEN`.  But it sounds like you agree  that best practice would be for the server to always return the same file handle for every nodeid.  Would you accept a PR to that effect?

**Comment by trapexit** [comment]
What would be the point of individual file handles if they are to be the same? You'd expect a FH to be part of lookup instead of open if that were the case. How is one supposed to differentiate between individual opens? F_GETFL and such would be impossible.

**Comment by asomers** [comment]
If a file system *must* discriminate between operations that come from different opens, then of course file handles are the way to do it.  But since FUSE clients can't perfectly track which operations are associated with which opens, clients would be advised not to rely upon it.

In the specific case of F_GETFL I don't see how file handles help.  They can't be used as a proxy for the lock owner.  The lock owner is the pid of the process that set the lock, not the process that opened the file.  And in order to be useful, file locks need to operate at the nodeid level, not the file handle level.

**Comment by trapexit** [comment]
> In the specific case of F_GETFL I don't see how file handles help. 

F_GETFL not F_GETLK. That's a per file handle/descriptor value is it not?

**Comment by asomers** [comment]
Sorry, I misread.  But what does `F_GETFL` have to do with FUSE?  Isn't it handled entirely within the kernel?

**Comment by trapexit** [comment]
I'm not talking about fcntl per se but using it as an example of the fact that some things are specific to file handles. If you want to have handle specific information (for whatever usage) then you have no other way but that handle to associate a request with that opened file. It's clear that the original design was 1 open file for 1 handle. Not one node. If that's not true (there are already issues with returning the handle with misc calls) then I'd agree there should be an overhaul of the documentation and API.

**Comment by Nikratio** [comment]
> "File descriptor" is not precisely the right term, but libfuse doesn't define the right term. What I'm referring to are concurrent opens of the same nodeid. While libfuse doesn't specify when the client should concurrently open the same nodeid multiple times, it's my impression that when `-o default_permissions` is not used, the client must FUSE_OPEN a nodeid for each and every `open(2)` call. Otherwise nothing would validate the permissions of the `open(2)`. So what's the correct term for each concurrent open of a nodeid?

I'd call that a "FUSE File Handle".

> Libfuse's documentation for `release` says this ("file descriptor" should probably say "file handle" instead):
> 
> ```
> * For every open() call there will be exactly one release() call
> * with the same flags and file descriptor.
> ```

Yes, this should be "file handle". 

> If "file descriptor" is supposed to be "file handle", then that clearly requires the client to track all file handles ever returned by `open`.

Yes. 

> And why would it track all of them if they were all required to be the same?

Nothing says that they're required to be the same, I think.

> Furthermore, the documentation for the `FUSE_WRITE_CACHE` bit says:
> 
> ```
> FUSE_WRITE_CACHE: delayed write from page cache, file handle is guessed
> ```
> 
> If the file handle is guessed when that bit is set, it implies that the file handle must not be guessed when that bit is cleared.

No. "If A then B" does *not* imply "if not(A) then not(B)". 

> From all of the above, it is pretty clear to me that for operations on an open file, the FUSE client must send the specific file handle returned by the server during `FUSE_OPEN`. 

Here it gets fuzzy. What do you mean with "operations on an open file"? It would be correct to say that when FUSE client (which is the kernel) wants to perform an operation on a FUSE file handle, it has to send this handle to the FUSE server. Is that what you mean? 

> But it sounds like you agree that best practice would be for the server to always return the same file handle for every nodeid. Would you accept a PR to that effect?

I still do not see how this follows from what you write above.

Would you be able to describe a specific scenario where a filesystem doesn't do this, and something undesired happens as a consequence?

**Comment by asomers** [comment]
> > Furthermore, the documentation for the `FUSE_WRITE_CACHE` bit says:
> > ```
> > FUSE_WRITE_CACHE: delayed write from page cache, file handle is guessed
> > ```
> > If the file handle is guessed when that bit is set, it implies that the file handle must not be guessed when that bit is cleared.
> 
> No. "If A then B" does _not_ imply "if not(A) then not(B)".

Why would the documentation specify that the file handle is guessed in this case, if it's always guessed?  It's specified in this case because it's the exception, not the rule.

> > From all of the above, it is pretty clear to me that for operations on an open file, the FUSE client must send the specific file handle returned by the server during `FUSE_OPEN`.
> 
> Here it gets fuzzy. What do you mean with "operations on an open file"? It would be correct to say that when FUSE client (which is the kernel) wants to perform an operation on a FUSE file handle, it has to send this handle to the FUSE server. Is that what you mean?

That's kind of tautological.  How about, "when the FUSE client wants to perform an operation on a file descriptor, it should use the file handle that the server returned when it opened that file descriptor"?

> 
> > But it sounds like you agree that best practice would be for the server to always return the same file handle for every nodeid. Would you accept a PR to that effect?
> 
> I still do not see how this follows from what you write above.
> 
> Would you be able to describe a specific scenario where a filesystem doesn't do this, and something undesired happens as a consequence?

Sure.  Most of the problems I can think of sound like a bad design choice on the part of the file system.  But that's because I think it's bad design to use multiple file handles for the same file.  Here's one scenario: suppose the file system is `O_APPEND`-aware.  The client opens a file with `O_APPEND`, then opens a second file without.  The user writes some data to offset 0 of the second file descriptor, but the client sends the file handle for the first file descriptor.  The server would wrongly assume that the data should be appended, rather than written to offset 0.

Here's another scenario.  Suppose the client opens the same file twice.  On the first time, the server sets `FOPEN_DIRECT_IO`, but on the second time it doesn't.  The user writes to the second file descriptor.  Because that file descriptor doesn't have `FOPEN_DIRECT_IO`, the kernel caches the writes.  But when it flushes the cache, it chooses the first file handle (perhaps the file handle guessing code always chooses the first file handle assigned to a vnode, or something).  Now the server gets a write that has `FUSE_WRITE_CACHE` set, but the file handle assumes that the file is non-cacheable, and an assertion fires.

Here's a third scenario.  Suppose the server handles authorization, and it uses some kind of mandatory access control.  Since MAC rules may change at any time, the server must authorize every single operation, not just open-like operations.  Suppose the client opens a file twice, and the server embeds different MAC labels into their file handles.  If the client guesses the wrong file handle for a read or write operation, then the server could make a wrong authorization decision for that operation.

**Comment by Nikratio** [comment]
> > > Furthermore, the documentation for the `FUSE_WRITE_CACHE` bit says:
> > > ```
> > > FUSE_WRITE_CACHE: delayed write from page cache, file handle is guessed
> > > ```
> > > If the file handle is guessed when that bit is set, it implies that the file handle must not be guessed when that bit is cleared.
> > 
> > 
> > No. "If A then B" does _not_ imply "if not(A) then not(B)".
> 
> Why would the documentation specify that the file handle is guessed in this case, if it's always guessed? It's specified in this case because it's the exception, not the rule.
>

It is not always guessed. The sentence above makes no statement at all about what file handles are passed to other operations.

> > > From all of the above, it is pretty clear to me that for operations on an open file, the FUSE client must send the specific file handle returned by the server during `FUSE_OPEN`.
> > 
> > 
> > Here it gets fuzzy. What do you mean with "operations on an open file"? It would be correct to say that when FUSE client (which is the kernel) wants to perform an operation on a FUSE file handle, it has to send this handle to the FUSE server. Is that what you mean?
> 
> That's kind of tautological. How about, "when the FUSE client wants to perform an operation on a file descriptor, it should use the file handle that the server returned when it opened that file descriptor"?

The concept of a file descriptor does not exist in the FUSE interface. So that sentence doesn't make any sense.

> > > But it sounds like you agree that best practice would be for the server to always return the same file handle for every nodeid. Would you accept a PR to that effect?
> > 
> > 
> > I still do not see how this follows from what you write above.
> > Would you be able to describe a specific scenario where a filesystem doesn't do this, and something undesired happens as a consequence?
> 
> Sure. Most of the problems I can think of sound like a bad design choice on the part of the file system. But that's because I think it's bad design to use multiple file handles for the same file. Here's one scenario: suppose the file system is `O_APPEND`-aware. The client opens a file with `O_APPEND`, then opens a second file without. The user writes some data to offset 0 of the second file descriptor, but the client sends the file handle for the first file descriptor. The server would wrongly assume that the data should be appended, rather than written to offset 0.

What do you mean with "the client"? The kernel will not do this, so this scenario cannot currently happen. It also seems that you are assuming a one to one mapping between FUSE file handles and file descriptors. There is no such unique mapping.

> Here's another scenario. Suppose the client opens the same file twice. 

Do you mean the kernel or userspace?

> On the first time, the server sets `FOPEN_DIRECT_IO`, but on the second time it doesn't. The user writes to the second file descriptor. 

I assume you mean userspace, because the kernel would write to a file handle.

> Because that file descriptor doesn't have `FOPEN_DIRECT_IO`, the kernel caches the writes.

File descriptors don't have this flag, this is a property of the FUSE file handle.

> But when it flushes the cache, it chooses the first file handle (perhaps the file handle guessing code always chooses the first file handle assigned to a vnode, or something). Now the server gets a write that has `FUSE_WRITE_CACHE` set, but the file handle assumes that the file is non-cacheable, and an assertion fires.

..and now I'm struggling to follow. Could you clarify the different "it" here? Also, how can a file handle assume something?

> Here's a third scenario. Suppose the server handles authorization, and it uses some kind of mandatory access control. Since MAC rules may change at any time, the server must authorize every single operation, not just open-like operations.

This would mean that read-ahead must be disabled, writeback caching must be disabled, and direct IO must be enabled for every file handle, right? Otherwise the filesystem won't see all requests.

> Suppose the client opens a file twice, and the server embeds different MAC labels into their file handles. If the client guesses the wrong file handle for a read or write operation, then the server could make a wrong authorization decision for that operation.

What do you mean with "guesses the wrong file handle"? The client (kernel) has two file handles, and it wants to write to one of them. I don't see how the concept or "wrong" or "right" enters here. I am assuming you refer to the mapping from userspace file descriptors to FUSE file handles, but it is still not clear to me how "wrongness" can enter here (especially with direct IO enabled).

**Comment by asomers** [comment]
Let me start over to try to make the problem more clear.

Problem
=======

1) The FUSE protocol uses server-side authorization (when
`-o default_permissions` is not used).

2) The FUSE protocol makes `FUSE_ACCESS` optional, and defaults to allow.  That
means a robust client cannot rely on `FUSE_ACCESS` to authorize open(2)
syscalls.

3) So the only remaining alternative for a client to authorize open(2) requests
is to use FUSE_OPEN, which must be sent for every open(2), or at least for
every open(2) that has unique credentials.  When `-o allow_other` is in use,
that means that a single file could generally have multiple concurrent file
handles.

4) open(2) returns a new file descriptor and FUSE_OPEN returns a new file
handle.  This effectively forces a 1:1 relationship between file descriptors
and file handles (although file descriptors with the same credentials can share
file handles).

5) The server may store credential-specific information in a file handle, so a
client cannot use FUSE_OPEN for authorization of a new open(2), then immediatly
FUSE_RELEASE the new file handle and reuse an old file handle for the new file
descriptor.

6) Unfortunately, most (all?) clients are unable to reliably associate a single
file descriptor with each FUSE operation that takes a file handle.  Linux
cannot do it when writing from the cache in writeback mode.  Nor can Linux do
it for most FUSE_SETATTR variations.  The problem is much worse on Solaris,
whose VFS enforces a distinction between file descriptors and vnodes, which
have a many-to-one relationship.  Solaris's problem was inherited by BSD, which
copied Solaris's VFS, and by OSX, which copied BSD's VFS.

7) It follows that a robust server cannot store credential-specific information in a file handle.

Solution
========

The solution is what I proposed in my OP.  Though another way to document the
restriction would be to advise server developers not to associate
credential-specific information with file handles.

FUSE Alternatives
=================

1) NFSv2 and NFSv3 don't have this problem because they use client-side
authorization.  I'm not well-versed enough to know how Kerberized NFSv4 deals
with the problem.

2) 9P doesn't have this problem because it has no equivalent of -o allow_other.
Each user mounts the server independently.

**Comment by Nikratio** [comment]
Thanks for the new write-up. I don't understand how your original solution proposal relates to this new description, but your second suggestion:

> Though another way to document the
> restriction would be to advise server developers not to associate
> credential-specific information with file handles.

is fine with me (I don't understand why people would think of doing that in the first place, but it doesn't hurt advising against doing that).

**Comment by asomers** [comment]
Ok, I'll submit a PR.  In an ideal world we could just fix the protocol by eliminating file handles or making FUSE_ACCESS mandatory or something like that.  But unfortunately we don't live in an ideal world and such a change would be painful.

**Comment by trapexit** [comment]
How would lseek or copy_file_range (with off_in = NULL) work without file handles? 

Also from an API perspective its pretty useful to embed state in the protocol and have it returned rather than manage it yourself.

**Comment by asomers** [comment]
Without file handles lseek would work just like it does for every other file system (including NFS and 9P); the file pointer would be a property of the file descriptor.  lseek would be a purely local operation, and all write operations would include an offset.

**Comment by Nikratio** [comment]
> In an ideal world we could just fix the protocol by eliminating file handles or making FUSE_ACCESS mandatory or something like that. B

Not sure if it matters, but I still do not understand what problem you are trying to fix with that. As far as I can tell, the previous examples that you gave weren't actual problems, and the new description that you gave amounts to "some people might try to associate credentials with file handles" - which is a bad idea no matter if FUSE_ACCESS is called more often, and doesn't seem enough of a justification to remove file handles. There's always a way to mis-use something, so with that philosophy we'd basically have to give up on software altogether.

**Comment by Nikratio** [comment]
> Without file handles lseek would work just like it does for every other file system (including NFS and 9P); the file pointer would be a property of the file descriptor. lseek would be a purely local operation, and all write operations would include an offset.

lseek() is a userspace operation that is never seen by libfuse, and all write operations that are sent to FUSE include an offset. 

I am sorry, but at this point I do not think following this issue is worth the amount of time that it takes. I'll try to take a look at a pull request if something comes out of this, but I'll probably refrain from further participation here.

**Comment by asomers** [comment]
The actual motivation for this issue is that I'm trying to make FreeBSD's FUSE driver work better.  But as I've explained, on non-Linux operating systems the relationship between file descriptors (and hence fuse file handles) and vnodes is much more tenuous.  So it's basically impossible for a non-Linux operating system to satisfy the 1:1 relationship between file descriptors and file handles that is implicit in FUSE's security model.  The three examples I gave _are_ real problems,  though unusual, because they require some legal-but-odd design choices on the part of the FS developer.

Also, regarding @trapexit's point, while all write operations do include _an_ offset, the file system must ignore that offset if it is to implement `O_APPEND`, which the docs say it must when writeback caching is disabled.
